mono/io-layer/wthreads.c

```diff
@@ -358,7 +352,11 @@ GetThreadPriority (gpointer handle)
 				  (gpointer *)&thread_handle);
 				  
 	if (ok == FALSE) {
-		return (THREAD_PRIORITY_NORMAL);
+		pthread_getschedparam (pthread_self(), &policy, &param);
+		if ((policy == SCHED_FIFO) || (policy == SCHED_RR)) 
+			return (((MonoInternalThread *)handle)->priority);
```

Comment:
Yes, for windows Get/Set need to be called with internal->handle (the value returned from the CreateThread() API). Which is interesting because under mono the CreateThread is not done until the thread is started. Therefore, internal->handle will be null. We could perhaps use that fact to use the stored value rather than invoking the API (Windows) or calling _wapi_GetThreadPriority() (POSIX). The same for the SetThreadPriority() call. Thus we can move priority out of MonoInternalThread and back into MonoThread. Note, the POSIX thread code uses the MonoInternalThread pointer as its handle and not internal->handle. So all lookups etc are done using it. Hence that's why there's a difference between Windows and POISIX. I am testing my theory at the moment.